### PR TITLE
Add AbstractCameraInterface and Picamera2 implementation

### PR DESCRIPTION
**Pull request by keiranbarr**
Added an AbstractCameraInterface for use in the Detector class, as well two concrete implementations--one for a USB webcam with OpenCV, the other with Picamera2.

### CODE REVIEW

File: src/detector/detector.py [hunk 1]
```diff
@@ -143,171 +100,33 @@ def supported_request_types(self) -> dict[type[MCastRequest], Callable[[dict], M
...
-            self._capture = None
-        self._capture_status.status = CaptureStatus.Status.STOPPED
-        return EmptyResponse()
-
-    '''
     # Camera
+    def internal_update_capture(self):
+        msg = self._camera_interface.internal_update_capture()
+        if msg:
+            self.add_status_message(msg)
```
Comment (vaughantnrc) [inline src/detector/detector.py:107]: I think severity might be omitted. But I also wonder if the overall status-reporting mechanism warrants a revisit. So I think it's okay to leave this as-is for now.

File: src/detector/detector.py [hunk 2]
```diff
@@ -143,171 +100,33 @@ def supported_request_types(self) -> dict[type[MCastRequest], Callable[[dict], M
...
+        return self._camera_interface.set_capture_device(**kwargs)
         
     def set_capture_properties(self, **kwargs) -> EmptyResponse:
-        self._camera_interface.set_capture_properties(**kwargs)
+        return self._camera_interface.set_capture_properties(**kwargs)
 
     def get_capture_device(self, **_kwargs) -> GetCaptureDeviceResponse:
-        # why underscored?
-        self._camera_interface.get_capture_device(**_kwargs)
+        # why underscored kwargs?
```
Comment (vaughantnrc) [inline src/detector/detector.py:116]: Indicates parameter is unused. The original reason for this was to satisfy a certain function signature requirement. I'm not sure with your changes if it still needs kwargs or not.

File: src/detector/detector_app.py [hunk 3]
```diff
@@ -33,18 +38,19 @@
...
     marker_interface: AbstractMarkerInterface
 
     with open(detector_configuration_filepath, 'r') as infile:
         detector_configuration_file_contents: str = infile.read()
         detector_configuration_dict = hjson.loads(detector_configuration_file_contents)
         detector_configuration = DetectorConfiguration(**detector_configuration_dict)
 
-    #camera_interface = Detector.USBWebcamWithOpenCV()
+    camera_interface = PiCamera()
+    # camera_interface = USBWebcamWithOpenCV(detector_configuration.camera_connection.usb_id)
```
Comment (vaughantnrc) [inline src/detector/detector_app.py:50]: I wonder if there's a cleaner way to pass an argument here to decide which camera to use

File: src/detector/implementations/abstract_camera_interface.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        pass
+
+    def get_capture_image(self, **kwargs) -> GetCaptureImageResponse:
+        pass
+
+    def start_capture(self, **kwargs) -> MCastResponse:
+        pass
+
+    def stop_capture(self, **kwargs) -> MCastResponse:
+        pass
```
Comment (vaughantnrc) [inline src/detector/implementations/abstract_camera_interface.py:42]: missing newline

File: src/detector/implementations/picamera2_implementation.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+from src.common import \
+    EmptyResponse, \
+    ErrorResponse, \
+    get_kwarg, \
+    MCastResponse
+from src.common.structures.capture_status import CaptureStatus
+
+from src.detector.implementations import AbstractCameraInterface
+
+from picamera2 import Picamera2, Preview
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:17]: Wouldn't this add a new requirement to the project?

File: src/detector/implementations/picamera2_implementation.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+    _capture: numpy.ndarray | None
+    _captured_image: numpy.ndarray | None
+
+    def __init__(self):
+        self._capture = None
+        self._captured_image = None
+        self._captured_timestamp_utc = datetime.datetime.min
+
+        self._capture_status = CaptureStatus
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:40]: Can you double check that you are creating an instance of `CaptureStatus` and not a reference to the class itself?

File: src/detector/implementations/picamera2_implementation.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        self._capture_status.status = CaptureStatus.Status.STOPPED
+
+        self._captured_timestamp_utc = datetime.datetime.min
+
+        self._capture_status = CaptureStatus
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:45]: Same as comment five lines above

File: src/detector/implementations/picamera2_implementation.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+import numpy
+import os
+from typing import Any, Callable
+
+logger = logging.getLogger(__name__)
+
+class PiCamera(AbstractCameraInterface):
+
+    _capture: numpy.ndarray | None
+    _captured_image: numpy.ndarray | None
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:33]: What is the difference between these?

Can you please add type hints for other class members?

File: src/detector/implementations/picamera2_implementation.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        return self._camera.capture_array()
+
+    def internal_update_capture(self):
+        self._captured_image = self._camera.capture_array()
+
+        if self._captured_image is None:
+            message: str = "Failed to grab frame."
+            self._status.capture_errors.append(message)
+            self._capture_status.status = CaptureStatus.Status.FAILURE
+            return ("error", message)
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:68]: Okay, I see. So this returns a `tuple` of str, str, not just a `str`. I suggest to add type hint to the function declaration, and that you make sure to unpack the variables properly when calling this function

File: src/detector/implementations/picamera2_implementation.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        """
+        :key request: SetCapturePropertiesRequest
+        """
+
+        request: SetCapturePropertiesRequest = get_kwarg(
+            kwargs=kwargs,
+            key="request",
+            arg_type=SetCapturePropertiesRequest)
+
+        if self._capture is not None:
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:103]: `_capture` vs `_camera`?

File: src/detector/implementations/picamera2_implementation.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                self._camera.controls.AeEnable = request.auto_exposure
+            # TODO: how to enforce values be entered in the proper range?
+            if request.exposure is not None:
+                self._camera.controls.ExposureValue = request.exposure
+            if request.brightness is not None:
+                self._camera.controls.Brightness = request.brightness
+            if request.contrast is not None:
+                self._camera.controls.Contrast = request.contrast
+            if request.sharpness is not None:
+                self._camera.controls.Sharpness = request.sharpness
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:125]: Are there other settings beyond these ones? I'm not asking for changes, just information

File: src/detector/implementations/picamera2_implementation.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        self._capture_status = CaptureStatus
+        self._capture_status.status = CaptureStatus.Status.STOPPED
+
+        self._captured_timestamp_utc = datetime.datetime.min
+
+        self._capture_status = CaptureStatus
+        self._capture_status.status = CaptureStatus.Status.STOPPED
+
+        # TODO: DEBUGGING
+        self._capture_status.status = CaptureStatus.Status.RUNNING
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:49]: still needed?

File: src/detector/implementations/picamera2_implementation.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+                resolution_x_px=int(self._camera.video_configuration.size[0]),
+                resolution_y_px=int(self._camera.video_configuration.size[1]),
+                fps=int(round(self._camera.video_configuration.controls.FrameRate)),
+                auto_exposure=bool(self._camera.controls.AeEnable),
+                exposure=int(self._camera.controls.ExposureValue),
+                brightness=self._camera.controls.Brightness,
+                contrast=self._camera.controls.Contrast,
+                sharpness=self._camera.controls.Sharpness)
+            return ret
+        # TODO: Get powerline_frequency_hz and backlight_compensation
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:147]: If picamera does not expose these, then just remove

File: src/detector/implementations/picamera2_implementation.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+            arg_type=GetCaptureImageRequest)
+
+        encoded_frame: bool
+        encoded_image_rgb_single_row: numpy.array
+        encoded, encoded_image_rgb_single_row = cv2.imencode(request.format, self._captured_image)
+        encoded_image_rgb_bytes: bytes = encoded_image_rgb_single_row.tobytes()
+        encoded_image_rgb_base64 = base64.b64encode(encoded_image_rgb_bytes)
+        return GetCaptureImageResponse(
+            format=request.format,
+            image_base64=encoded_image_rgb_base64)
```
Comment (vaughantnrc) [inline src/detector/implementations/picamera2_implementation.py:166]: I wonder if it makes sense to put this (or part of this) in superclass

File: src/detector/implementations/usb_webcam_implementation.py [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+import datetime
+import logging
+import numpy
+import os
+from typing import Any, Callable
+
+logger = logging.getLogger(__name__)
+
+# TODO: update type hints
+class USBWebcamWithOpenCV(AbstractCameraInterface):
```
Comment (vaughantnrc) [inline src/detector/implementations/usb_webcam_implementation.py:28]: I will assume that this class is more or less verbatim for what existed previously, so I won't go through it in detail

**Review by vaughantnrc — CHANGES_REQUESTED** [review]

File: src/detector/implementations/picamera2_implementation.py (hunk 5 above)
Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:17]: Picamera has to be installed with apt, not pip in my experience

File: src/detector/implementations/picamera2_implementation.py (hunk 8 above)
Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:33]: For the original code for the USB webcam there was a distinction between _capture and _captured_image. I copied that structure for the Pi camera implementation but can change it as it is now unnecessary.

I have type hints for capture_status and captured_timestamp_utc in the parent abstract class. Should I move/copy them to the child class?

File: src/detector/implementations/picamera2_implementation.py (hunk 11 above)
Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:125]: Yes there are, a full list can be found [here](https://datasheets.raspberrypi.com/camera/picamera2-manual.pdf#camera-controls)

File: src/detector/implementations/picamera2_implementation.py (hunk 12 above)
Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:49]: I am not sure, I copied it from the original code. I'm not entirely sure what its purpose is.

File: src/detector/detector.py (hunk 2 above)
Comment (keiranbarr) [reply src/detector/detector.py:116]: The function in the detector class still needs kwargs but I suppose I could omit passing it to the corresponding function in the abstract interface. Or should I just leave it alone?

File: src/detector/implementations/picamera2_implementation.py (hunk 12 above)
Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:49]: Interestingly, it seems obsolete in the USB webcam implementation, but is still necessary in the pi camera one.

**Comment by keiranbarr** [comment]
Please see my new commit. I have addressed most of these comments, and left responses to those that I am unsure of how (or if) to address.

File: src/detector/detector.py (hunk 2 above)
Comment (vaughantnrc) [reply src/detector/detector.py:116]: Wouldn't other functions still need kwargs? I think it's safest just to leave this alone for now (but remove the comment).

File: src/detector/implementations/picamera2_implementation.py (hunk 5 above)
Comment (vaughantnrc) [reply src/detector/implementations/picamera2_implementation.py]: Would I be correct in thinking that this dependency will not be resolvable on Windows?

File: src/detector/implementations/picamera2_implementation.py (hunk 11 above)
Comment (vaughantnrc) [reply src/detector/implementations/picamera2_implementation.py:121]: Thanks. There are certainly a number of interesting settings there. It looks to me like the exposure used in OpenCV and the exposure used in picamera possibly use different conventions. I think we can leave this as-is for now, but we will for sure later want to have different input data structures depending on whether the OpenCV or PiCamera driver is being used.

File: src/detector/implementations/picamera2_implementation.py (hunk 5 above)
Comment (vaughantnrc) [reply src/detector/implementations/picamera2_implementation.py]: If so, then perhaps the comment above will be important to address before merge

https://github.com/PerkLab/MCSTrack/pull/30#discussion_r1616316556

File: src/detector/implementations/picamera2_implementation.py (hunk 12 above)
Comment (vaughantnrc) [reply src/detector/implementations/picamera2_implementation.py:45]: What happens if you remove the line of code? Does starting the camera work as intended?

File: src/detector/implementations/picamera2_implementation.py (hunk 11 above)
Comment (vaughantnrc) [reply src/detector/implementations/picamera2_implementation.py:121]: Defer this to later:  #32

File: src/detector/implementations/picamera2_implementation.py (hunk 12 above)
Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:45]: The camera starts and an initial frame is shown but it does not update.

Comment (vaughantnrc) [reply src/detector/implementations/picamera2_implementation.py:45]: and are you referring to the GUI?

Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py:45]: My apologies, I should be more clear. If I remove the line then the camera still starts and connects to the GUI just fine, but after the initial frame appears in the GUI the image is never updated. And yes, I am not changing my usage of the .start() command specific to the Picamera2 python library, nor the AbstractCameraInterface.start_capture() method.

Comment (keiranbarr) [reply src/detector/implementations/picamera2_implementation.py]: This is now fixed on my machine with my latest commit.

File: src/detector/detector_app.py [hunk 16]
```diff
@@ -24,27 +28,35 @@
...
+    camera_interface: AbstractCameraInterface
     marker_interface: AbstractMarkerInterface
 
     with open(detector_configuration_filepath, 'r') as infile:
         detector_configuration_file_contents: str = infile.read()
         detector_configuration_dict = hjson.loads(detector_configuration_file_contents)
         detector_configuration = DetectorConfiguration(**detector_configuration_dict)
 
-    #camera_interface = Detector.USBWebcamWithOpenCV()
+    if detector_configuration.camera_implementation == "opencv":
```
Comment (vaughantnrc) [inline src/detector/detector_app.py:46]: these strings to become constants

File: src/detector/implementations/usb_webcam_implementation.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+
+    _capture: cv2.VideoCapture | None
+    _captured_image: numpy.ndarray | None
+    _captured_timestamp_utc: datetime.datetime
+    _capture_device_id: str | int
+    _capture_status: CaptureStatus  # internal bookkeeping
+
+    def __init__(self, _capture_device_id):
+        # move to camera class
```
Comment (vaughantnrc) [inline src/detector/implementations/usb_webcam_implementation.py:36]: remove comment

File: src/detector/implementations/usb_webcam_implementation.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+    def __del__(self):
+        if self._capture is not None:
+            self._capture.release()
+
+    def _detect_os_and_open_video(self,capture_device_id):
+        if os.name == 'nt':
+            return cv2.VideoCapture(capture_device_id, cv2.CAP_DSHOW)
+        else:
+            return cv2.VideoCapture(capture_device_id)
```
Comment (vaughantnrc) [inline src/detector/implementations/usb_webcam_implementation.py:53]: Add comment explaining why these are different

**Review by vaughantnrc — APPROVED** [review]
Minor changes, then please merge.

**Pull request merged**